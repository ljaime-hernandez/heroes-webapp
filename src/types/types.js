
// the types contain a string in this scenario just so they have information
// but the content is irrelevant for this example
export const types = {
    login: '[auth] login',
    logout: '[auth] logout'
}
