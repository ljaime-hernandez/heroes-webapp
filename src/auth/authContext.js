import { createContext } from "react";

// the createContext method returns a component, thats why
// we declare it with a capital letter at first, we will
// use this context which will cover all the routes of our
// webapp to transfer certain information across the components,
// in this case to handle private and public routes within the
// page
export const AuthContext = createContext();