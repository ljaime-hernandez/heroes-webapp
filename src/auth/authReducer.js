import { types } from "../types/types";

// the reducer should be a pure function in a manner on which everything
// should be resolved within it, the reducer should not exit to check on
// additional functions but only to work in itself and give an outcome no
// matter what arguments it receive, in this case our default return is 
// an empty object which is a default value for the state
export const authReducer = (state = {}, action) => {

    switch (action.type) {
        case types.login:
                return {
                    ...action.payload,
                    logged: true
                }

        case types.logout:
                return {
                    logged: false
                }

        default:
            return state;
    }

}