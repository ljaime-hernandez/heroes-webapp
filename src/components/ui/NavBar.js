import React, { useContext } from 'react';
import { Link, NavLink, useNavigate } from 'react-router-dom';
import { AuthContext } from '../../auth/authContext';
import { types } from '../../types/types';

export const Navbar = () => {

    // we use the useNavigate method from the react-router-dom library to assign its content on a 
    // constant which we will later use for a click event
    const navigate = useNavigate();
    // we call the useContext method which is transferring the user information with all the routers
    // on the webpage, im destructuring it from the AuthContext component and then i retrieve
    // its name to render it properly on the navigation bar
    const {user, dispatch} = useContext(AuthContext);
    const {name} = user;

    const handleLogout = () => {

        // once the button is clicked, we will send the user to the login page, rendering its contents.
        // the 'replace: true' is used so if the user tries to return to the previous page, he is unable
        // to do so, we do this so once the user is logged out, he cannot come back until he login again,
        // not very specific on this example but conditionally makes sense on a session context
        const action = {
            type: types.logout,
        }

        dispatch(action);
        
        
        navigate('/login', {
            replace: true,
        });
    }

    return (
        <nav className="navbar navbar-expand-sm navbar-dark bg-dark">
            
            <Link 
                className="navbar-brand" 
                to="/"
            >
                Comics
            </Link>

            <div className="navbar-collapse">
                <div className="navbar-nav">
                    {/*the condition on the NavLink component will check if the url where the user is
                       located is the same as the element being rendered on the NavBar, if it is then the
                       class will highlight the text contained, letting the user know which comic publisher
                       he is checking now */}
                    <NavLink 
                        className={({isActive}) => "nav-item nav-link" + (isActive ? ' active' : '')} 
                        to="/marvel"
                    >
                        Marvel
                    </NavLink>
                    {/*the condition on the NavLink component will check if the url where the user is
                       located is the same as the element being rendered on the NavBar, if it is then the
                       class will highlight the text contained, letting the user know which comic publisher
                       he is checking now */}               
                    <NavLink 
                        className={({isActive}) => "nav-item nav-link" + (isActive ? ' active' : '')} 
                        to="/dc"
                    >
                        DC
                    </NavLink>

                    {/*the condition on the NavLink component will check if the url where the user is
                       located is the same as the element being rendered on the NavBar, if it is then the
                       class will highlight the text contained, letting the user know which comic publisher
                       he is checking now */}  
                    <NavLink 
                        className={({isActive}) => "nav-item nav-link" + (isActive ? ' active' : '')} 
                        to="/search"
                    >
                        Search
                    </NavLink>
                </div>
            </div>

            <div className="navbar-collapse collapse w-100 order-3 dual-collapse2 d-flex justify-content-end">
                <ul className="navbar-nav ml-auto">

                    <span
                        className='nav-item nav-link text-info'
                    >
                        {name}
                    </span>

                    <button 
                        className="nav-item nav-link btn" 
                        onClick={handleLogout}
                    >
                        Logout
                    </button>
                </ul>
            </div>
        </nav>
    )
}