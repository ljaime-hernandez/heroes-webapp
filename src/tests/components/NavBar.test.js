import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { AuthContext } from '../../auth/authContext';
import { Navbar } from '../../components/ui/NavBar';
import { types } from '../../types/types';

/* to run this test:
1. run the 'npm install' command from the heroes-webapp folder 
2. run the 'npm install --save-dev enzyme' command (if you have not done so)
3. run the 'npm install --save-dev enzyme-to-json' command (if you have not done so)
4. run the 'npm install --save-dev @wojtekmaj/enzyme-adapter-react-17 --legacy-peer-deps' command (if you are using React 17 as i do)
5. run the 'npm install --save-dev @testing-library/react-hooks' command (if you have not done so)
6. make sure the setupTests.js file include the enzyme, enzyme-to-json and the react adapter libraries
7. run the command 'npm run test'
8. to have a clearer view of this single js test file, press p. then type the file name 'Navbar.test.js'
*/

let dispatch = jest.fn();
let mockNavigate = jest.fn();
// same as with the SearchScreen component, if we dont create a navigate mock function then the mount 
// method would give us an error. As the dispatch belongs to our custom reducer, we dont get errors
// when trying to render it, but we still need to create a mock from it as the functionality is not
// scripted in this component, but its just a function passed as a parameter along with the user from the
// context.
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate
}));


describe('tests on NavBar component', () => {


    test('should render properly', () => {
        
        const contextValue = {
            user: {
                logged: true,
                name: 'Pedro'
            }}
            
        // we need both the user context and the router context for us to render the 
        // NavBar component, so both our higher-order components need to be
        // implemented for the screenshot to work 
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <AuthContext.Provider value={contextValue}>
                    <Navbar/>
                </AuthContext.Provider>
            </MemoryRouter>
            )
        
        expect(wrapper).toMatchSnapshot();
        // in a normal render for the NavBar component, we would display the user name by the
        // logout button, the element containing the user name is called text-info, so we 
        // use the find method to get it, then we retrieve its text and match it with this
        // test to confirm is working accordingly
        expect(wrapper.find('.text-info').text().trim()).toBe('Pedro');
    })

    test('should call handle logout function, use navigate and dispatch', () => {
        
        const contextValue = {
            dispatch,
            user: {
                logged: true,
                name: 'Pedro'
            }}
            
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <AuthContext.Provider value={contextValue}>
                    <Navbar/>
                </AuthContext.Provider>
            </MemoryRouter>
            )

            // we can use both a simulate or a prop method to call on the click event,
            // in here i do the prop method just to variate, the parentheses at the end of 
            // the method is for it to call the function inside its property
            wrapper.find('button').prop('onClick')();
            // on the actual NavBar component, when the click event is called, then a
            // function called handleLogout is going to trigger a series of actions, 
            // it will first retrieve the types.logout from the types file, then pass
            // it as a parameter to the dispatch function from the user reducer.
            // We did a mock of the dispatch method as the dispatch itself is tested
            // in the authReducer tests, so we just need to confirm it was called
            // accordingly in here with the respective type
            expect(contextValue.dispatch).toHaveBeenCalledWith({'type': types.logout});
            // same as in the previous test, the navigate method is inside the handleLogout
            // function, we dont need to test the navigate functionality but we just want to
            // confirm it was called with the right parameter
            expect(mockNavigate).toHaveBeenCalledWith('/login', {replace: true});
    })
})