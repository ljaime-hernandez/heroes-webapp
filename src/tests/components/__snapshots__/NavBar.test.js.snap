// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests on NavBar component should render properly 1`] = `
<nav
  className="navbar navbar-expand-sm navbar-dark bg-dark"
>
  <a
    className="navbar-brand"
    href="/"
    onClick={[Function]}
  >
    Comics
  </a>
  <div
    className="navbar-collapse"
  >
    <div
      className="navbar-nav"
    >
      <a
        className="nav-item nav-link"
        href="/marvel"
        onClick={[Function]}
      >
        Marvel
      </a>
      <a
        className="nav-item nav-link"
        href="/dc"
        onClick={[Function]}
      >
        DC
      </a>
      <a
        aria-current="page"
        className="nav-item nav-link active"
        href="/search"
        onClick={[Function]}
      >
        Search
      </a>
    </div>
  </div>
  <div
    className="navbar-collapse collapse w-100 order-3 dual-collapse2 d-flex justify-content-end"
  >
    <ul
      className="navbar-nav ml-auto"
    >
      <span
        className="nav-item nav-link text-info"
      >
        Pedro
      </span>
      <button
        className="nav-item nav-link btn"
        onClick={[Function]}
      >
        Logout
      </button>
    </ul>
  </div>
</nav>
`;
